### ISSUE TITLE
Add rule to validate color functions

### ISSUE THREAD

**Issue by Mottie** [stylelint/stylelint#2816]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

RGB(A) and HSL(A) color definition warnings are missing.

> Which rule, if any, is this issue related to?

A new rule is required.

> What CSS is needed to reproduce this issue?

```css
a {
  color: rgb(200%, 20%); /* min 3 definition */
  color: rgb(0, 0, 0, 0, 0, 0, 0); /* max 3 definitions */
  color: rgba(0); /* min 4 definitions */
  color: rgba(0, 0, 0, 0, 0, 0, 0); /* max 4 definitions */
  color: rgba(0, 0, 0, 100%); /* alpha channel can not be a percentage */
}

b {
  color: hsl(0); /* min 3 definitions */
  color: hsl(100%); /* hue can not be a percentage (number, deg, rad or turn) */
  color: hsl(0, 0, 0, 0, 0, 0, 0); /* max 3 definitions; */
  color: hsl(0, 50, 50); /* saturation & lightness must be percentages */
  color: hsla(0); /* min 4 definitions */
  color: hsla(0, 0, 0, 0, 0, 0, 0); /* max 4 definitions */
  color: hsla(0, 0%, 0%, 0%); /* alpha channel can not be a percentage */
}
```

> What stylelint configuration is needed to reproduce this issue?

Default configuration used by the [online demo](https://stylelint.io/demo/).

> Which version of stylelint are you using?

`8.0.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Online demo.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No, this is standard syntax.

> What did you expect to happen?

For RGB(A) colors:
* WARNING/ERROR: There must be 3 rgb and 4 rgba definitions.
* WARNING?: RGB colors can be either an integer or a percentage. Out of bound values do not cause errors.
* WARNING/ERROR: Alpha channel setting can not be a percentage.
* WARNING/ERROR: Don't mix integers and percentages ([ref](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#RGB_example)).

For HSL(A) colors:
* WARNING/ERROR: There must be 3 hsl and 4 hsla definitions.
* WARNING/ERROR: Hue (H) definition must be an integer. Out of bound values do not cause errors.
* WARNING/ERROR: Saturation/Lightness must be percentages. Out of bound values do not cause errors.
* WARNING/ERROR: Alpha channel setting can not be a percentage.

Update:

[According to MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#Examples), you can also use a functional syntax in Firefox v52+ (not yet supported in other browsers):

* Space-separated function parameters rather than commas.
* Percentages for alpha values.
* Angles for the hue component in `hsl()` colors.

```css
/* these are valid definitions (no commas, use slash) */
rgb(255 0 153 / 1)
rgb(255 0 153 / 100%)
hsl(270 60% 50% / .15)
hsl(270 60% 50% / 15%)
/* valid use of deg, rad or turn */
hsl(270deg, 60%, 70%)
hsl(4.71239rad, 60%, 70%)
hsl(.75turn, 60%, 70%)
```

> What actually happened (e.g. what warnings or errors you are getting)?

Everything looks legit - no warnings/errors.

**Comment by evilebottnawi** [comment]
@Mottie thanks for issue!

**Comment by evilebottnawi** [comment]
I think best name `color-function-no-invalid`.
@Mottie Maybe we can include in this rule other color function? Can your to keep in mind and examine their?

**Comment by Mottie** [comment]
I was going to leave the rule name up to you all... I just updated the issue with more valid examples [according to MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#Examples).

**Comment by ntwb** [comment]
Thanks for the detailed issue @Mottie, sounds good to me, I've updated the labels for this issue 👍 

Would you be interested in taking a shot at creating this new `color-function-no-invalid` rule?

**Comment by jeddy3** [comment]
In the [VISON.md](https://github.com/stylelint/stylelint/blob/master/VISION.md) we have:

> Invalid code is best handled by emerging dedicated tools e.g. csstree - a language parser with syntax validation. As a stop-gap, while these tools mature, provide invalid code rules for the simplest of cases.

I'm not sure this is the "simplest of cases". Perhaps it's better to add this type of validation to a tool like `csstree`?

**Comment by jeddy3** [comment]
@Mottie FYI, it looks like the [stylelint-csstree-validator](https://github.com/csstree/stylelint-validator) plugin has come along a lot recently. I tested it against your examples and it was able to warn against them all except the "alpha channel can not be a percentage" ones.

I think we can close this issue for now, and continue with the approach from [VISON.md](https://github.com/stylelint/stylelint/blob/master/VISION.md). Let's help push the syntax validation in csstree forward. We can promote the plugin in issues (and revise our documentation to emphasise it once it matures).

**Comment by Mottie** [comment]
Actually the "alpha channel can not be a percentage" is [valid in Firefox](https://developer.mozilla.org/en-US/docs/Web/CSS/color#Browser_compatibility), but not supported in any other browser. Anyway thanks! I'll take a look at csstree.
